File: app/src/main/java/social/entourage/android/base/EntourageLinkMovementMethod.java
package social.entourage.android.base;

import android.text.Layout;
import android.text.Spannable;
import android.text.method.LinkMovementMethod;
import android.text.method.MovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.URLSpan;
import android.view.MotionEvent;
import android.widget.TextView;

import social.entourage.android.api.tape.Events;
import social.entourage.android.tools.BusProvider;

/**
 * Created by Mihai Ionescu on 11/04/2018.
 */
public class EntourageLinkMovementMethod extends LinkMovementMethod {

    private static EntourageLinkMovementMethod instance;

    public static MovementMethod getInstance() {
        if (instance == null) {
            instance = new EntourageLinkMovementMethod();
        }
        return instance;
    }

    @Override
    public boolean onTouchEvent(final TextView widget, final Spannable buffer, final MotionEvent event) {
        final int action = event.getAction();
        if (action == MotionEvent.ACTION_DOWN || action == MotionEvent.ACTION_UP) {

            final int x = (int) event.getX() - widget.getTotalPaddingLeft() + widget.getScrollX();
            final int y = (int) event.getY() - widget.getTotalPaddingTop() + widget.getScrollY();
            final Layout layout = widget.getLayout();
            final int line = layout.getLineForVertical(y);
            final int off = layout.getOffsetForHorizontal(line, x);
            final ClickableSpan[] links = buffer.getSpans(off, off, ClickableSpan.class);
            if (links.length != 0) {
                // to avoid double handling, we handle the link only on down
                if (action == MotionEvent.ACTION_DOWN) {
                    if (links[0] instanceof  URLSpan) {
                        String url = ((URLSpan) links[0]).getURL();
                        BusProvider.INSTANCE.getInstance().post(new Events.OnShowURLEvent(url));
                    }

                }
                return true;
            }
        }
        return super.onTouchEvent(widget, buffer, event);
    }

}

File: app/src/main/java/social/entourage/android/base/EntourageLinkMovementMethod.kt
package social.entourage.android.base

import android.text.Spannable
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.URLSpan
import android.view.MotionEvent
import android.widget.TextView
import social.entourage.android.api.tape.Events.OnShowURLEvent
import social.entourage.android.tools.BusProvider

/**
 * Created by Mihai Ionescu on 11/04/2018.
 */
object EntourageLinkMovementMethod : LinkMovementMethod() {
    override fun onTouchEvent(widget: TextView, buffer: Spannable, event: MotionEvent): Boolean {
        val action = event.action
        if (action == MotionEvent.ACTION_DOWN || action == MotionEvent.ACTION_UP) {
            val x = event.x.toInt() - widget.totalPaddingLeft + widget.scrollX
            val y = event.y.toInt() - widget.totalPaddingTop + widget.scrollY
            val layout = widget.layout
            val line = layout.getLineForVertical(y)
            val off = layout.getOffsetForHorizontal(line, x.toFloat())
            val links = buffer.getSpans(off, off, ClickableSpan::class.java)
            if (links.isNotEmpty()) {
                // to avoid double handling, we handle the link only on down
                if (action == MotionEvent.ACTION_DOWN) {
                    if (links[0] is URLSpan) {
                        val url = (links[0] as URLSpan).url
                        BusProvider.instance.post(OnShowURLEvent(url))
                    }
                }
                return true
            }
        }
        return super.onTouchEvent(widget, buffer, event)
    }
}